SharpMUSH.Implementation/Functions/MailFunctions.cs

```diff
@@ -1,64 +1,603 @@
-﻿using SharpMUSH.Library.Attributes;
+﻿using SharpMUSH.Implementation.Commands.MailCommand;
+using SharpMUSH.Library;
+using SharpMUSH.Library.Attributes;
 using SharpMUSH.Library.Definitions;
+using SharpMUSH.Library.DiscriminatedUnions;
+using SharpMUSH.Library.Extensions;
+using SharpMUSH.Library.Models;
 using SharpMUSH.Library.ParserInterfaces;
+using SharpMUSH.Library.Queries.Database;
+using SharpMUSH.Library.Services.Interfaces;
 
 namespace SharpMUSH.Implementation.Functions;
 
 public partial class Functions
 {
+	/// <summary>
+	/// Parse message specification (e.g. "123" or "INBOX:5") into folder and message index
+	/// </summary>
+	private static async ValueTask<(string folder, int messageIndex)> ParseMessageSpec(
+		IMUSHCodeParser parser,
+		AnySharpObject player,
+		string messageSpec)
+	{
+		var parts = messageSpec.Split(':', 2);
+		string folder;
+		int messageIndex;
+
+		if (parts.Length == 2 && !string.IsNullOrWhiteSpace(parts[0]))
+		{
+			// Format: "folder:number"
+			folder = parts[0].Trim().ToUpper();
+			if (!int.TryParse(parts[1].Trim(), out messageIndex) || messageIndex < 1)
+			{
+				return (folder, -1);
+			}
+		}
+		else
+		{
+			// Format: just "number" - use current folder
+			folder = await MessageListHelper.CurrentMailFolder(parser, ObjectDataService!, player);
+			if (!int.TryParse(messageSpec.Trim(), out messageIndex) || messageIndex < 1)
+			{
+				return (folder, -1);
+			}
+		}
+
+		// Convert from 1-indexed to 0-indexed
+		return (folder, messageIndex - 1);
+	}
+
+	/// <summary>
+	/// Retrieve mail message by folder and index
+	/// </summary>
+	private static async ValueTask<SharpMail?> GetMailMessage(
+		AnySharpObject player,
+		string folder,
+		int messageIndex)
+	{
+		return await Mediator!.Send(new GetMailQuery(player.AsPlayer, messageIndex, folder));
+	}
+
+	/// <summary>
+	/// Helper to parse target player and message spec from function arguments.
+	/// Returns null if permission denied or player not found.
+	/// </summary>
+	private static async ValueTask<(AnySharpObject? player, string? messageSpec)> ParsePlayerAndMessageArgs(
+		IMUSHCodeParser parser,
+		AnySharpObject executor,
+		Dictionary<string, CallState> args)
+	{
+		if (args.Count == 1)
+		{
+			return (executor, args["0"].Message!.ToPlainText()!);
+		}
+
+		// Two arguments - must be wizard to view other player's mail
+		if (!executor.IsGod() && !await executor.IsWizard())
+		{
+			return (null, null);
+		}
+
+		var playerArg = args["0"].Message!.ToPlainText()!;
+		var locateResult = await LocateService!.LocateAndNotifyIfInvalid(
+			parser, executor, executor, playerArg, LocateFlags.PlayersPreference);
+
+		if (locateResult.IsError || locateResult.IsNone)
+		{
+			return (null, null);
+		}
```

Comment:
Consider how to improve the return value or this method in general so that the semantics of IsError/IsNone and the permission check are actually kept and the data isn't lost.